src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,169 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addToTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of range!");
+        }
+        if (index == size) {
+            addToTail(value);
+        } else {
+            addByIndex(value, index);
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T val : list) {
+            add(val);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexInRange(index);
+        return getCurrentNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexInRange(index);
+        Node<T> currentNode = getCurrentNode(index);
+        T prevValue = currentNode.value;
+        currentNode.value = value;
+        return prevValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexInRange(index);
+        Node<T> currentNode = getCurrentNode(index);
+        T deletedNode = currentNode.value;
+        deleteNode(currentNode);
+        return deletedNode;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        int counter = 0;
+        while (counter < size) {
+            if (object != null && object.equals(currentNode.value) || object == currentNode.value) {
+                deleteNode(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+            counter++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addToTail(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        newNode.prev = tail;
+        tail = newNode;
+        if (newNode.prev == null) {
```

Comment:
let's change this check, for example you can use `size == 0` to check if we are adding the first node